homeassistant/components/imeon_inverter/coordinator.py

```diff
@@ -0,0 +1,154 @@
+    thus it's own data set. This allows this integration to handle as many
+    inverters as possible in parallel.
+    """
+
+    _HUBs: dict[Any, InverterCoordinator] = {}
+
+    # Implement methods to fetch and update data
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entry: InverterConfigEntry,
+        user_input: dict[str, Any] | None = None,
+    ) -> None:
+        """Initialize data update coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            # Name of the data. For logging purposes.
+            name=HUBNAME,
+            # Polling interval. Will only be polled if there are subscribers.
+            update_interval=timedelta(minutes=1),
+            always_update=True,
+            config_entry=entry,
+        )
+
+        if user_input is None:
+            raise ValueError("user_input cannot be None")
+
+        self.api = Inverter(user_input["address"])  # API calls
+        self.username = user_input["username"]
+        self.password = user_input["password"]
+        self.friendly_name = entry.title
+
+        # unique ID
+        self.__id = entry.entry_id
+        InverterCoordinator._HUBs[str(self.__id)] = self
+
+        # Store request data
+        self.data = {}
+        self.first_call = True
+
+    def update(self, user_input: dict[str, Any]) -> None:
+        """Update HUB data based on user input."""
+        self.api = Inverter(user_input["address"])
+        self.username = user_input["username"]
+        self.password = user_input["password"]
+        self.first_call = True
+
+    @property
+    def id(self):
+        """Getter for id."""
+        return self.__id
+
+    @staticmethod
+    def get_from_id(id) -> InverterCoordinator:
+        """Getter for InverterCoordinator."""
+        try:
+            return InverterCoordinator._HUBs[str(id)]
+        except IndexError:
+            raise IndexError(f"Incorrect HUB ID ({id!s}) .") from None
+
+    def store_data(self, entity_dict):
+        """Store in data for entities to use."""
+        for key in entity_dict:
+            if key != "timeline":
+                val = entity_dict[key]
+                for sub_key, sub_val in val.items():
+                    self.data[key + "_" + sub_key] = sub_val
+            else:  # Timeline is a list not a dict
+                self.data[key] = entity_dict[key]
+
+    def get_storage(self) -> dict[str, Any]:
+        """Get all the value from the API."""
+        return {
+            "battery": self.api.battery,
+            "grid": self.api.grid,
+            "pv": self.api.pv,
+            "input": self.api.input,
+            "output": self.api.output,
+            "meter": self.api.meter,
+            "temp": self.api.temp,
+            "monitoring": self.api.monitoring,
+            "manager": self.api.manager,
+            "inverter": self.api.inverter,
+            "timeline": self.api.timeline,
+            "smartload": {},
```

Comment:
why are you rewrapping those? maybe just return self.api from update. Return None and access the api from the coordinator in your entities